### ISSUE TITLE
Feature Request: Linux support

### ISSUE THREAD

**Issue by funkymonkeymonk** [felixhageloh/uebersicht#185]
Dropping this issue in as a place for others to help/ advise.

I'm reading through the code now to try to understand how uebersicht places widgets on the desktop and to figure out if there is anyway to do this on a linux DE as well. I would appreciate any advice on how to approach this problem and also feedback on if this is something that the maintainers would like to see done, and would be willing to merge in if it works.

**Comment by hawkrives** [comment]
What I can say is that there are two logical parts to Übersicht: the server, and the OS X app that renders the server on the desktop.

The server is written in JS, and run through node.js; there shouldn't really be any problems related to that, with the potential exception of the Übersicht-specific extensions (under `window.uebersicht`, like the geolocation apis.)

The OS X app, though, is entirely cocoa. It's what renders the server's web page beneath the icons, and what adds the menulet to the menu bar. That would need to be entirely re-created for Linux, and would ideally support whatever today's status icon specification is, so that it could work across DEs (KDE, GNOME, XFCE, etc.)

In short, I guess, have fun? I, at least, would love to know if you get it working on Linux.

**Comment by funkymonkeymonk** [comment]
Cool. I believe that each DE is implementing status icons differently. The closest to a standard is that everyone seems to have plugins for Unity's app indicators, which is likely where I would start. https://unity.ubuntu.com/projects/appindicators/ 

As for rewriting the display client, I'm wondering if the most effective path to take with this wouldn't be attempting a rewrite of the Objective C into swift, then creating an interface class between the Cocoa calls and the caller, and subbing the Cocoa called for gtk in linux.

**Comment by felixhageloh** [comment]
A Linux version would be nice! Conky was actually one of biggest the inspirations for Übersicht. 

As @hawkrives pointed out, there is two parts to the app: the node backend and the cocoa frontend which mostly renders a transparent full screen web view over the desktop + a status bar menu to interact with the app.

There is actually not that much generic code in the cocoa app. It is mostly just dealing with OS X specific APIs for showing windows, storing user preferences, detecting screens etc. Rendering widgets is entirely done by the web view. So I wonder how much would be reusable even if it was written in Swift. 

I think the smallest iteration to get something working would be a Linux app that starts the node server task and renders a web view that connects to localhost (you can specify a port when starting the server). Widgets are loaded from a directory path you also provide to the server. The only pitfall there is that it uses a native node module (fsevents) to load and watch for widget files, so that would require a small rewrite using a generic file watcher.

In any case, shout if you have mote questions!

**Comment by ljsherlock** [comment]
Any update on the @funkymonkeymonk ? I just discovered Übersicht because I missed having Conky on my Mac work machine. Now I'm hooked.

**Comment by bauen1** [comment]
If someone finds a way to do the equivialent of
```
[webView setValue:@YES forKey:@"drawsTransparentBackground"];
```
in linux on a webkit2-gtk WebKitWebView, then I might be able to hack something together.

**Comment by hawkrives** [comment]
@bauen1 Eh, "just" render the webview behind the icons, then use übersicht to draw a wallpaper widget.

(Note: I have no idea what I'm talking about.)

**Comment by bauen1** [comment]
I had some time again, and implemented a "webkit2gtk browser" https://gitlab.com/bauen1/uebersicht
I'm still having some issues with building the client, but it should be possible as the frontend of a linux version.

**Comment by mstr3336** [comment]
@bauen1 , what steps would I need to take to get this working with the node backend?

#286 seems to suggest that the backend is coupled to OSX's `fsevents` for the time being, but it seems like @felixhageloh  is open to cross platform alternatives
